benchmarks/python/core.py

```diff
@@ -259,40 +260,107 @@ def run_benchmark(
         outputs: Output of the target function
     """
 
-    assert device in [
+    # Check that the device is `cuda` or `host:{compile/steady/dynamic}`.
+    assert device.split(":")[0] in [
         "cuda",
         "host",
-    ], f"Unsupported device type: {device}. Use one of cuda/host."
+    ], f'Unsupported device type: {device.split(":")[0]}. Use one of cuda/host.'
 
-    if device == "host":
+    host_bench_mode = None
+
+    # Store warmup rounds locally to modify for host:steady/dynamic cases.
+    global warmup_rounds
+    warmup_rounds = BENCHMARK_CONFIG["warmup_rounds"]
+
+    if device.split(":")[0] == "host":
+        # Host benchmarking expects a fusion function to generate fusion definitions everytime FusionCache is reset.
         assert fusion_fn is not None and benchmark_fn is None
 
-    nvf_benchmark = NVFBenchmark(benchmark, device=device)
+        # Reset the FusionCache to avoid any inadvertent fusion execution from affecting measurements
+        FusionCache.reset()
+
+        # device = 'host:compile', 'host:steady', 'host:dyanamic'
+        # Set the host_bench_mode -- The 3 modes require different setup calls.
+        host_bench_mode = device.split(":")[-1]
+        device = device.split(":")[0]  # device = 'host'
+
+        # Set the warmup rounds if required for `steady/dynamic` host latency measurement.
+        if (
+            host_bench_mode in ["steady", "dynamic"]
+            and BENCHMARK_CONFIG["warmup_rounds"] == 0
+        ):
+            # By default, warmup_rounds=1. If BENCHMARK_CONFIG['warmup_rounds'] == 0 through --benchmark-warmup-rounds, raise a warning that it was ignored.
+            warnings.warn(
+                "--benchmark-warmup-rounds=0 is ignored for host:steady/dynamic benchmarking. Setting warmup_rounds=1."
+            )
+            warmup_rounds = 1
 
-    def host_benchmark_fn(inputs, fd):
-        # Set the fd variable used to query the profile object
-        nvf_benchmark.fd = fd
-        return fd.execute(inputs, profile=True)
+    """
+    Setup function: This is called before each benchmarking round. This function is used to:
+    1. Clear L2 cache.
+    2. For host latency benchmarks, the 3 modes use different setups.
+        a) 'compile': FusionCache is reset at every round to measure the first time overhead before instantiating fd.
+        b) 'steady': Nothing additional is required. The warmup round avoids including the first time overhead in the measurements.
+        c) 'dynamic': We maintain a global counter to track which input is executed. Once all the inputs have been executed once,
+        the FusionCache is reset again and we execute fd for the first input to avoid including the first time compile overhead in the dynamic measurement.
+    """
+
+    # Counter used in `dynamic` host latency benchmarking, unused in other cases.
+    global counter
```

Comment:
same here

Reply:
This needs to be declared global to modify within different calls of the `setup` function.

Reply:
IIRC, you don't need to have `global` here since you have `global` inside the `setup` function (at line 336), but I'm not an active user of Python anymore, so maybe I'm wrong.

Reply:
It is needed here since the declaration of `counter` is within the `run_benchmark` function itself, so by default, the scope of `counter=0` without the global declaration will be the outer function and the inner `setup` function cannot modify this local variable of the outer function. If it was defined at the module level, outside of this function, we would not need this `global`